### ISSUE TITLE
[v1.4.4.1]  Imports fail & throw a ReferenceError on client load. (Project contains a .babelrc file)

### ISSUE THREAD

**Issue by cfnelson** [meteor/meteor#8600]
After updating to Meteor v1.4.4.1 from v1.4.3.2 the app would successfully build/load but on client load the below error would be thrown.

```
Uncaught ReferenceError: React is not defined
    at index.js:13
    at maybeReady (meteor.js?hash=27829e9…:809)
    at HTMLDocument.loadingCompleted (meteor.js?hash=27829e9…:821)
```
The import that caused the above ReferenceError is `import React from 'react';` which worked fine in Meteor v1.4.3.2. The  v1.4.3.2 project's `.babelrc` file is below: 

```
{
  "presets": ["meteor", "env", "react", "stage-2"],
  "plugins": ["transform-es2015-modules-commonjs"]
}
```

I was able to resolve the above ReferenceError for Meteor v1.4.4.1 with the below steps:

`meteor reset` & `rm -rf node_modules` & `meteor npm install`
`meteor npm install --save-dev babel-plugin-transform-es2015-modules-reify`

I updated the `.babelrc` to use reify directly.
```
{
  "presets": ["meteor", "env", "react", "stage-2"],
  "plugins": ["transform-es2015-modules-reify"]
}
```
After the changes above I was able to start the Meteor app without any issues.

Our .babelrc file is used for running tests, it may be our initial config was incorrect but I tried removing `env`, `react` from the preset and a few other preset options (nothing helped). 

The changes I made to fix the `ReferenceError` broke our `mocha` tests. To resolve this I installed `reify` directly as a dependency `meteor npm install --save reify` and passed reify as an option to mocha's `--compiler` command.

Potentially related post in the [forums](https://forums.meteor.com/t/ant-design-breaks-on-meteor-1-4-4/35676/4)

I am curious to know what the exact issue/cause was and what the ideal solution/fix is then the one I was able to come up with? Hopefully others can benefit from the above if they experience a similar issue.

**Comment by benjamn** [comment]
Where is the `import React from 'react'` declaration in your code? Can you provide a small reproduction repo demonstrating the problem?

**Comment by benjamn** [comment]
Can you try running `meteor update babel-compiler` (or even `meteor update --all-packages`)? This problem may be the same as #8595, which was just fixed.

**Comment by cfnelson** [comment]
@benjamn Thanks for the quick response!

I should have mentioned where the React import was located. It's in `imports/client/index.js`. It's at the first line of the file. The index.js file is handling the initial client startup. e.g - rendering the <Root/> for the app.
```
render(
    <Root/>,
    document.getElementById('app'),
  ); 
```

I will set aside some time tomorrow and try the above suggestions as well as attempt reproduce the issue in a small repo.

If this helps, the repo started with `babel-compiler@6.14.1`, after updating to Meteor v1.4.4.1 the package was updated to: `babel-compiler@6.18.1`.

**Comment by benjamn** [comment]
The version you want is `babel-compiler@6.18.2`.

**Comment by cfnelson** [comment]
@benjamn Sorry for the delay. I have created a small reproduction repo [here](https://github.com/cfnelson/meteor-issue-8600) for the `Uncaught ReferenceError: React is not defined`. I also attempted updating to `babel-compiler@6.18.2` as well as  `meteor update --all-packages` without any luck. You will see these commits in the reproduction repo.

Hope this is helpful, let me know if I should make any further changes.

**Comment by talha-asad** [comment]
I am having a slightly different and weird issue since upgrading to 1.4.4.1

```js
if (Meteor.isServer) {
  import { isValidImage, transformExpressImage } from '/imports/server/utilities/file';
}


AdImageMediaStore = new UploadFS.store.GridFS({
  collection: AdMedia,
  name: 'AdImageMedia',
  chunkSize: 1024 * 1024, // 1MB
  filter: new UploadFS.Filter({
    minSize: 1,
    maxSize: 10 * 1024 * 1024, // 10MB
    contentTypes: ['image/*'],
    extensions: ['gif', 'jpg', 'jpeg', 'png', 'bmp']
  }),
  onValidate(file) {
    if (typeof file.resize === 'object') {
      const tempFilePath = UploadFS.getTempFilePath(file._id);
      return isValidImage(tempFilePath); // is undefined here on the server ?? :S
    }
  },
  transformWrite(from, to, fileId, file) {
    if (typeof file.resize === 'object') {
      transformExpressImage(from, to, fileId, file); // is undefined here on the server ?? :S
    }
  },
  permissions: new UploadFS.StorePermissions({
    insert(userId, doc) {
      return userId;
    },
    update(userId, doc) {
      return userId === doc.userId;
    },
    remove(userId, doc) {
      return userId === doc.userId;
    }
  })
});
```

`isValidImage` and `transformExpressImage` are undefined in those functions? This works without an issue on Meteor 1.4.3

Any clue?

**Comment by hwillson** [comment]
@talha-asad https://github.com/meteor/meteor/issues/8546#issuecomment-290664999 should help.

**Comment by benjamn** [comment]
@talha-asad Nested `import` declarations now declare the imported symbols with block scope, per https://github.com/meteor/meteor/issues/8021, so `isValidImage` behaves more like a `let` variable than a `var` variable.

You can either move `AdImageMediaStore` inside the `if (Meteor.isServer) { ... }` block (if appropriate), or externalize the variables explicitly:
```js
let isValidImage, transformExpressImage;
if (Meteor.isServer) {
  import { isValidImage as a, transformExpressImage as b } from '/imports/server/utilities/file';
  isValidImage = a;
  transformExpressImage = b;
}
```

**Comment by benjamn** [comment]
@cfnelson Short answer: you should remove the `babel-plugin-transform-es2015-modules-commonjs` plugin from your `.babelrc` file [here](https://github.com/cfnelson/meteor-issue-8600/blob/6a864caf5710b5185f901542c0c4607641ba9a92/.babelrc#L3).

Explanation: in Meteor 1.4.4, any custom Babel plugins in `.babelrc` files now run to completion before any other plugins, thanks to [this commit](https://github.com/meteor/babel/commit/c30fe307fc1ce184f3aa8aa7712c87688bceb090). This change was made to simplify reasoning about Babel plugin order, since by default plugins are "interleaved" for performance reasons, meaning plugin ordering is only preserved with respect to individual AST nodes. In particular, a plugin that visits the root `Program` node will always run before one that visits any other kind of node, regardless of how you try to order them. Letting custom plugins finish before presets begin gives you more control over how your code is transformed, though it can have noticeable consequences (like this one).

Since `babel-preset-meteor` already compiles `import` and `export` syntax, your `babel-plugin-transform-es2015-modules-commonjs` was always unnecessary, but now it's running first, and appears to be interacting badly with other plugins, because it rewrites code like
```js
import React from "react";
React.createClass(...)
```
to
```js
var _react = require("react");
_react.React.createClass(...)
```
but not all references to `React` are correctly rewritten, perhaps because the `babel-plugin-transform-es2015-modules-commonjs` is running before the JSX transform, so there are new references to `React` created later.

In other words, you shouldn't be running `babel-plugin-transform-es2015-modules-commonjs` this early in your plugin pipeline, or running it at all.

Related: https://github.com/meteor/meteor/issues/8578#issuecomment-294257682

**Comment by cfnelson** [comment]
@benjamn Thanks for the detailed explanation! I can confirm the repo works with the removal of `transform-es2015-modules-commonjs` from the plugins.

**Comment by talha-asad** [comment]
@benjamn That's a really ugly solution. I don't see why you would want imports to be block scope, Reading up on #8021, The author should not name the import variable the same again and again. The code below isn't working anymore, even though I did use block imports.
Require works perfectly but then it's not the latest way of doing stuff.

**ufs_collection.js**
```js
AdImageMediaStore = new UploadFS.store.GridFS({
  collection: AdMedia,
  name: 'AdImageMedia',
  chunkSize: 1024 * 1024, // 1MB
  filter: new UploadFS.Filter({
    minSize: 1,
    maxSize: 10 * 1024 * 1024, // 10MB
    contentTypes: ['image/*'],
    extensions: ['gif', 'jpg', 'jpeg', 'png', 'bmp']
  }),
  onValidate(file) {
    import { isValidImage } from '/imports/server/utilities/file';
    if (typeof file.resize === 'object') {
      const tempFilePath = UploadFS.getTempFilePath(file._id);
      return isValidImage(tempFilePath);
    }
  },
  transformWrite(from, to, fileId, file) {
    import { transformExpressImage } from '/imports/server/utilities/file';
    if (typeof file.resize === 'object') {
      transformExpressImage(from, to, fileId, file);
    }
  },
  permissions: new UploadFS.StorePermissions({
    insert(userId, doc) {
      return userId;
    },
    update(userId, doc) {
      return userId === doc.userId;
    },
    remove(userId, doc) {
      return userId === doc.userId;
    }
  })
});
```

**/imports/server/utilities/file.js**
```js
import { gm } from 'gm';
import { Future } from 'fibers/future';
import { _ } from 'lodash';

export const isValidImage = function(filePath) {
  let future = new Future(); // Not accessible here, when ufs function calls it.
  gm(filePath).identify((err, data) => {
    if (err)
      return future.throw(new Meteor.Error('not-an-image', 'The file is not an image'));
    future.return(data);
  });
  return future.wait();
};

export const transformExpressImage = function(from, to, fileId, file) {
  gm(from) // Not accessible here, when ufs function calls it.
    .autoOrient()
    .noProfile()
    .resize(file.resize.width, file.resize.height, '!')
    .gravity('Center')
    .quality(82)
    .stream().pipe(to);
};
```

Why is the above not working, Isn't the file import even with block scope suppose to be available to `isValidImage` and `transformExpressImage` ?

**Comment by benjamn** [comment]
@talha-asad The `Future` constructor is the default export of `fibers/future`, so you want `import Future from 'fibers/future'` (no braces) there.

You know enough now to make your own decisions about what's clean and what's ugly. Inlining the imports where you're using them (what you're doing) is much better than what I suggested. I was just trying to illustrate why your original code is broken now.

This issue now closed, and unrelated to your questions, so let's stop this conversation. I would recommend creating a new issue, but I have to be honest with you that block scoping of imports is here to stay, so it's probably not worth your time to open an issue about that.

**Comment by talha-asad** [comment]
@benjamn are the braces making it block scope? If yes, then is this part of the spec? I just don't want to adapt to practices that are not future-proof.
